examples/amm/lib.rs

```diff
@@ -0,0 +1,725 @@
+    #[derive(
+        Default, Copy, PartialEq, Eq, Debug, Clone, scale::Decode, scale::Encode,
+    )]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub struct PoolDetails {
+        total_token1: Token1,
+        total_token2: Token2,
+        total_lp_tokens: LPTokens,
+        fee: Balance,
+    }
+
+    /// Used to describe if token1 was swapped for Token2 or if Token2 was swapped for
+    /// Token1
+    #[derive(Debug, PartialEq, Eq, scale::Encode, scale::Decode)]
+    #[cfg_attr(feature = "std", derive(scale_info::TypeInfo))]
+    pub enum SwapType {
+        Token1ForToken2,
+        Token2ForToken1,
+    }
+
+    /// Used to describe how much of each token was minted
+    #[ink(event)]
+    pub struct TokensMinted {
+        #[ink(topic)]
+        account: AccountId,
+        #[ink(topic)]
+        token1_amount: Token1,
+        #[ink(topic)]
+        token2_amount: Token2,
+    }
+
+    /// Used to describe how much of each token was deposited
+    /// amount_token1 - how many of token1 was deposited by the liquidity provider
+    /// amount_token2 - how many of token2 was deposited by the liquidity provider
+    /// amount_lp - how many lp_tokens were given in exchange for both tokens
+    #[ink(event)]
+    pub struct LiquidityProvided {
+        #[ink(topic)]
+        account: AccountId,
+        #[ink(topic)]
+        token1_amount: Token1,
+        #[ink(topic)]
+        token2_amount: Token2,
+        #[ink(topic)]
+        amount_lp: LPTokens,
+    }
+
+    /// Used to describe how much of each token was withdrawn from the pool
+    /// amount_token1 - how many of token1 was withdrawn by the liquidity provider
+    /// amount_token2 - how many of token2 was withdrawn by the liquidity provider
+    /// amount_lp - how many lp_tokens were burned in exchange for both tokens
+    #[ink(event)]
+    pub struct LiquidityWithdrawn {
+        #[ink(topic)]
+        account: AccountId,
+        #[ink(topic)]
+        token1_amount: Token1,
+        #[ink(topic)]
+        token2_amount: Token2,
+        #[ink(topic)]
+        amount_lp: LPTokens,
+    }
+
+    /// Used to describe how many tokens were swapped
+    /// deposit_amount - the amount of Token1 or Token2 that was given.
+    /// swap_amount - the amount of Token1 or Token2 that was received. If Token1 was deposited
+    /// swap_type - Used to describe which token was deposited
+    #[ink(event)]
+    pub struct TokenSwapped {
+        #[ink(topic)]
+        account: AccountId,
+        #[ink(topic)]
+        deposit_amount: Token1,
+        #[ink(topic)]
+        swap_amount: Token2,
+        #[ink(topic)]
+        swap_type: SwapType,
+    }
+
+    impl Amm {
+        // ==============================================================================
+        // 1. Constructor
+        // ==============================================================================
+        /// The constructor takes two arguments: fee, and precision. fee it used to
```

Comment:
```suggestion
        /// fee it used to
```

We can see this from the function signature 😉